packages/react/dropdown-menu/src/DropdownMenu.stories.tsx

```diff
@@ -312,6 +313,215 @@ export const WithLabels = () => (
   </div>
 );
 
+export const WithAsDialogTrigger = () => {
+  const dropdownTriggerRef = React.useRef<React.ElementRef<typeof DropdownMenuTrigger>>(null);
+  const dropdownTriggerRef2 = React.useRef<React.ElementRef<typeof DropdownMenuTrigger>>(null);
+  const isDialogOpenRef = React.useRef(false);
+
+  function handleModalDialogClose(event: Event) {
+    // focus dropdown trigger for accessibility so user doesn't lose their place in the document
+    dropdownTriggerRef.current?.focus();
+    event.preventDefault();
+  }
+
+  function handleNonModalDialogClose(event: Event) {
+    // focus dropdown trigger for accessibility so user doesn't lose their place in the document
+    dropdownTriggerRef2.current?.focus();
+    event.preventDefault();
+    isDialogOpenRef.current = false;
+  }
+
+  return (
+    <div
+      style={{
+        display: 'flex',
+        flexDirection: 'column',
+        alignItems: 'center',
+        justifyContent: 'center',
+        height: '100vh',
+      }}
+    >
+      <h1>Modal</h1>
+      <Dialog.Root>
+        <DropdownMenu>
+          <DropdownMenuTrigger className={triggerClass} ref={dropdownTriggerRef}>
+            Open
+          </DropdownMenuTrigger>
+
+          <DropdownMenuContent className={contentClass} sideOffset={5}>
+            <Dialog.Trigger className={itemClass} as={DropdownMenuItem}>
+              Delete
+            </Dialog.Trigger>
+            <DropdownMenuItem className={itemClass}>Test</DropdownMenuItem>
+            <DropdownMenuArrow />
+          </DropdownMenuContent>
+        </DropdownMenu>
+
+        <Dialog.Content className={dialogClass} onCloseAutoFocus={handleModalDialogClose}>
+          <Dialog.Title>Are you sure?</Dialog.Title>
+          <Dialog.Close>Close</Dialog.Close>
+        </Dialog.Content>
+      </Dialog.Root>
+
+      <h1>Non-modal</h1>
+      <Dialog.Root modal={false}>
+        <DropdownMenu modal={false}>
+          <DropdownMenuTrigger className={triggerClass} ref={dropdownTriggerRef2}>
+            Open
+          </DropdownMenuTrigger>
+
+          <DropdownMenuContent
+            className={contentClass}
+            sideOffset={5}
+            onCloseAutoFocus={(event) => {
+              // prevent focusing dropdown trigger when it closes from a dialog trigger
+              if (isDialogOpenRef.current) event.preventDefault();
+            }}
```

Comment:
@benoitgrelard We shouldn't really have to do this because the `DropdownMenu` already has logic to prevent its trigger from focusing if there was a focus outside of it. However, [the `setTimeout` in `FocusScope` for `onCloseAutoFocus`](
https://github.com/radix-ui/primitives/blob/main/packages/react/focus-scope/src/FocusScope.tsx#L133) is re-ordering events and breaking the expected behaviour here.

We have discussed removing the `setTimeout` from `FocusScope` given that v17 has been out for a while and we are a new modern library but we're reluctant to make that change as part of this PR (not sure of the knock on effects in other primz).

We're going to leave this in the story for now to demonstrate the workaround but it would be good to discuss this in more detail with you when you return.

Reply:
Because the `modal` scenario traps focus in the dialog so when it tries to focus the dropdown trigger on `onCloseAutoFocus` it can't.